jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java

```diff
@@ -0,0 +1,362 @@
+     * </pre>
+     * <p>
+     * The the {@code /admin/*} constraint would be consider most specific and a request to {@code /admin/file} would
+     * have {@link Authorization#SPECIFIC_ROLE} from the {@code /admin/*} constraint and
+     * {@link Transport#SECURE} inherited from the {@code /*} constraint.  For more examples see
+     * {@link SecurityHandler.PathMapped}.
+     * </p>
+     * <p>Note that this combination is not equivalent to the combination done by the EE servlet specification.</p>
+     * @param name The name to use for the combined constraint
+     * @param leastSpecific Constraint to combine
+     * @param mostSpecific Constraint to combine
+     * @return the combined constraint.
+     */
+    static Constraint combine(String name, Constraint leastSpecific, Constraint mostSpecific)
+    {
+        if (leastSpecific == null)
+            return mostSpecific == null ? ALLOWED : mostSpecific;
+        if (mostSpecific == null)
+            return leastSpecific;
+
+        return from(
+            name,
+            mostSpecific.getTransport() == Transport.INHERIT ? leastSpecific.getTransport() : mostSpecific.getTransport(),
+            mostSpecific.getAuthorization() == Authorization.INHERIT ? leastSpecific.getAuthorization() : mostSpecific.getAuthorization(),
+            mostSpecific.getAuthorization() == Authorization.INHERIT ? leastSpecific.getRoles() : mostSpecific.getRoles());
+    }
+
+    static Constraint from(String... roles)
+    {
+        return from(null, Authorization.SPECIFIC_ROLE, roles);
+    }
+
+    static Constraint from(String name, Transport transport)
+    {
+        return from(name, transport, null, null);
+    }
+
+    static Constraint from(String name, Authorization authorization, String... roles)
+    {
+        return from(name, Transport.INHERIT, authorization, (roles == null || roles.length == 0)
+            ? Collections.emptySet()
+            : new HashSet<>(Arrays.stream(roles).toList()));
+    }
+
+    static Constraint from(Transport transport, Authorization authorization, Set<String> roles)
+    {
+        return from(null, transport, authorization, roles);
+    }
+
+    static Constraint from(String name, Transport transport, Authorization authorization, Set<String> roles)
+    {
+        Set<String> roleSet = roles == null || roles.isEmpty()
+            ? Collections.emptySet()
+            : Collections.unmodifiableSet(roles);
+
+        Authorization auth = authorization == null
+            ? (roleSet.isEmpty() ? Authorization.INHERIT : Authorization.SPECIFIC_ROLE)
+            : authorization;
+
+        if (!roleSet.isEmpty() && auth != Authorization.SPECIFIC_ROLE)
+            throw new IllegalStateException("Constraint with roles must be SPECIFIC_ROLE, not " + auth);
+
+        return new Constraint()
+        {
+            @Override
+            public String getName()
+            {
+                return name == null ? "unnamed@%x".formatted(hashCode()) : name;
```

Comment:
Given that roles and authorization are normalized from `null` to a non-null value in `from()`, can we do the same for name and transport? Seems weird that we initialize some, but then have ternary expressions in the getters for others.
Doing so will also normalize `toString()` which now would print a `null` transport instead of `INHERIT`.